src/library_promise.js

```diff
@@ -0,0 +1,162 @@
+      // Allocate space for the result value and initialize it to NULL.
+      var resultPtr = stackAlloc(POINTER_SIZE);
+      {{{ makeSetValue('resultPtr', 0, '0', '*') }}};
+      try {
+        var result =
+            {{{ makeDynCall('ippp', 'callback') }}}(resultPtr, userData, value);
+        var resultVal = {{{ makeGetValue('resultPtr', 0, '*') }}};
+      } catch (e) {
+        // If the thrown value is potentially a valid pointer, use it as the
+        // rejection reason. Otherwise use a null pointer as the reason.
+#if MEMORY64
+        if (typeof e !== 'bigint') {
+          throw 0n;
+        }
+#else
+        if (typeof e !== 'number') {
+          throw 0;
+        }
+#endif
+        throw e;
+      } finally {
+        // Thrown errors will reject the promise, but at least we will restore
+        // the stack first.
+        stackRestore(stack);
+      }
+      if (result == {{{ cDefine('EM_PROMISE_FULFILL') }}}) {
+        return resultVal;
+      } else if (result == {{{ cDefine('EM_PROMISE_MATCH') }}}) {
+        return getPromise(resultVal);
+      } else if (result == {{{ cDefine('EM_PROMISE_MATCH_RELEASE') }}}) {
+        var ret = getPromise(resultVal);
+        _emscripten_promise_destroy(resultVal);
+        return ret;
+      } else if (result == {{{ cDefine('EM_PROMISE_REJECT') }}}) {
+        throw resultVal;
+      }
+#if ASSERTIONS
+      abort("unexpected promise callback result " + result);
+#endif
+    };
+  },
+
+  emscripten_promise_then__deps: ['$promiseMap',
+                                  '$getPromise',
+                                  '$makePromiseCallback'],
+  emscripten_promise_then__sig: 'ppppp',
+  emscripten_promise_then: function(id,
+                                    onFulfilled,
+                                    onRejected,
+                                    userData) {
+#if RUNTIME_DEBUG
+    dbg('emscripten_promise_then: ' + id);
+#endif
+    var promise = getPromise(id);
+    var newId = promiseMap.allocate({
+      promise: promise.then(makePromiseCallback(onFulfilled, userData),
+                            makePromiseCallback(onRejected, userData))
+    });
+#if RUNTIME_DEBUG
+    dbg('create: ' + newId);
+#endif
+    return newId;
+  },
+
+  emscripten_promise_all__deps: ['$promiseMap', '$getPromise', '$POINTER_SIZE'],
+  emscripten_promise_all__sig: 'pppp',
+  emscripten_promise_all: function(idBuf, resultBuf, size) {
+    var promises = [];
+    for (var i = 0; i < size; i++) {
+      var id = {{{ makeGetValue('idBuf', 'i*POINTER_SIZE', 'i32') }}};
+      promises[i] = getPromise(id);
+    }
+#if RUNTIME_DEBUG
+    dbg('emscripten_promise_all: ' + promises);
+#endif
+    var id = promiseMap.allocate({
+      promise: Promise.all(promises).then((results) => {
+        if (resultBuf) {
+          for (var i = 0; i < size; i++) {
+            var result = results[i];
+            {{{ makeSetValue('resultBuf', 'i*POINTER_SIZE', 'result', '*') }}};
```

Comment:
I think we prefer to use the compile time POINTER_SIZE.  Its pretty odd but in other places we do `i*${Runtime.POINTER_SIZE}` here

Reply:
(in backticks).